packages/eslint-plugin/src/util.ts

```diff
@@ -14,46 +15,81 @@ export const createRule: <TOptions extends readonly unknown[], TMessageIds exten
 );
 
 export function getTypesPackageForDeclarationFile(file: string) {
-  if (!file.endsWith(".d.ts")) {
+  if (!isDeclarationPath(file)) {
     return undefined;
   }
+  const info = findTypesPackage(file);
+  return info?.realName;
+}
 
-  const match = file.match(/types\/([^\/]+)\//)?.[1];
-  if (!match) {
-    return undefined;
+export function commentsMatching(
+  sourceFile: Readonly<SourceCode>,
+  regex: RegExp,
+  f: (match: string, c: TSESTree.Comment) => void
+): void {
+  for (const comment of sourceFile.ast.comments) {
+    const m = comment.value.match(regex);
+    if (m) f(m[1], comment);
   }
+}
 
-  return unmangleScopedPackage(match) ?? match;
+export interface TypesPackageInfo {
+  dir: string;
+  packageJson: PackageJSON;
+  realName: string;
 }
 
-export function isMainFile(fileName: string, allowNested: boolean) {
-  // Linter may be run with cwd of the package. We want `index.d.ts` but not `submodule/index.d.ts` to match.
-  if (fileName === "index.d.ts") {
-    return true;
-  }
+export interface PackageJSON {
+  name: string;
+  version: string;
+  owners: string[];
+  devDependencies?: Record<string, string | undefined>;
+}
 
-  if (basename(fileName) !== "index.d.ts") {
-    return false;
-  }
+// TODO(jakebailey): pull this helper out to util package?
+function isTypesPackage(packageJson: Partial<PackageJSON>): boolean {
+  return (
+    typeof packageJson.name === "string" &&
+    packageJson.name.startsWith("@types/") &&
+    typeof packageJson.version === "string" &&
+    Array.isArray(packageJson.owners)
+  );
+}
+
+export function findTypesPackage(file: string): TypesPackageInfo | undefined {
+  file = path.resolve(file);
+  const root = path.parse(file).root;
 
-  let parent = dirname(fileName);
-  // May be a directory for an older version, e.g. `v0`.
-  // Note a types redirect `foo/ts3.1` should not have its own header.
-  if (allowNested && /^v(0\.)?\d+$/.test(basename(parent))) {
-    parent = dirname(parent);
+  for (let dir = path.dirname(file); dir !== root; dir = path.dirname(dir)) {
```

Comment:
This parent traversal is actually very cheap; doing this for every single TS file in DT only took .2s on my machine, while a version of this code that got fancy with caching the info at each directory took .5s.

Reply:
Probably I should pull this parent walk out into its own helper; I use it twiceish.